### PR TITLE
descs: batch descriptor access

### PR DESCRIPTION
**Pull request by postamar**
This PR effectively amortizes the costs of bulk GRANT/REVOKE operations, but the main reason why I'm doing this now is that I need this for the changes I want to make to `scbuild`.

The change has been broken down into 3 commits for ease of review.

### CODE REVIEW

**Review by ajwerner — APPROVED** [review]
Well, this rocks and was not a big deal. My review isn't that deep, I'll be honest.

File: pkg/sql/catalog/descs/descriptor.go [hunk 1]
```diff
@@ -26,128 +26,207 @@ import (
...
+func (q *byIDLookupContext) lookupLeased(id descpb.ID) (catalog.Descriptor, error) {
+	if q.flags.AvoidLeased || q.flags.RequireMutable || lease.TestingTableLeasesAreDisabled() {
+		return nil, nil
+	}
+	// If we have already read all of the descriptors, use it as a negative
+	// cache to short-circuit a lookup we know will be doomed to fail.
+	//
+	// TODO(ajwerner): More generally leverage this set of kv descriptors on
+	// the resolution path.
+	if q.tc.kv.idDefinitelyDoesNotExist(id) {
```
[the remark this answers is not in the document]
Comment (postamar) [reply pkg/sql/catalog/descs/descriptor.go:225]: I just moved this around from where it was before, so I'm going to leave it there, at least until we revisit how `kv` works and caches stuff.

Really, this negative cache seems of dubious utility here: how often do you have the ID of a descriptor but the descriptor doesn't actually exist? And even then, so what?

**Comment by postamar** [comment]
Thanks for the review. What helped make this into not big of a deal was getting rid of `catalogkv`, I think.

**Comment by ajwerner** [comment]
> I think.

I know this is the case. Without separating the building and validation, this wouldn't have been possible.

**Comment by ajwerner** [comment]
bors r- 

looks like merge skew 

```
# github.com/cockroachdb/cockroach/pkg/migration/migrations
pkg/migration/migrations/remove_invalid_database_privileges.go:92:71: cannot use idAndVersion.id (type catid.DescID) as type *kv.Txn in argument to descriptors.GetMutableDescriptorByID
pkg/migration/migrations/remove_invalid_database_privileges.go:92:71: cannot use txn (type *kv.Txn) as type catid.DescID in argument to descriptors.GetMutableDescriptorByID
```

**Comment by postamar** [comment]
Indeed, thanks for catching it. Fixed.

bors r+

**Pull request merged**